### PR TITLE
Add Siddhi persistence on S3 feature

### PR DESCRIPTION
**Pull request by sahandilshan**
## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Test environment
- JDK version - 1.8
- aws.sdk.version - 2.10.1
- apache.httpclient.version - 4.5.10

### CODE REVIEW

File: runner/components/io.siddhi.distribution.core/src/main/java/io/siddhi/distribution/core/persistence/S3PersistenceStore.java [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+            log.error("Error occurred while trying to compress the snapshot. Failed to " +
+                    "persist revision: " + revision + " of Siddhi app: " + siddhiAppName);
+            return;
+        }
+        RequestBody requestBody = RequestBody.fromBytes(compressedSnapshot);
+        PutObjectResponse putObjectResponse =
+                s3Client.putObject(PutObjectRequest.builder().bucket(bucketName).key(revision)
+                        .build(), requestBody);
+        cleanOldRevisions();
+        log.debug("object has been uploaded to the bucket successfully, ETag: " + putObjectResponse.eTag());
```
Comment (mohanvive) [inline runner/components/io.siddhi.distribution.core/src/main/java/io/siddhi/distribution/core/persistence/S3PersistenceStore.java]: You have to put a isDebugEnabled flag here.

File: runner/components/io.siddhi.distribution.core/src/main/java/io/siddhi/distribution/core/persistence/S3PersistenceStore.java [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+    @Override
+    public void setProperties(Map properties) {
+        Map configurationMap = (Map) properties.get(PersistenceConstants.STATE_PERSISTENCE_CONFIGS);
+        Object numberOfRevisionsObject = properties.get(PersistenceConstants.STATE_PERSISTENCE_REVISIONS_TO_KEEP);
+        String regionId;
+
+        if (!(numberOfRevisionsObject instanceof Integer)) { //it also check if it's Null or not
+            numberOfRevisionsToSave = PersistenceConstants.DEFAULT_REVISION_NO;
+            log.warn("Number of revisions to keep is not set or invalid. Default value will be used.");
+
```
Comment (mohanvive) [inline runner/components/io.siddhi.distribution.core/src/main/java/io/siddhi/distribution/core/persistence/S3PersistenceStore.java]: unnecessary line space.

File: runner/components/io.siddhi.distribution.core/src/main/java/io/siddhi/distribution/core/persistence/S3PersistenceStore.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+        } else {
+            numberOfRevisionsToSave = Integer.parseInt(String.valueOf(numberOfRevisionsObject));
+        }
+
+        if (configurationMap != null) {
+            Object regionObject = configurationMap.get(PersistenceConstants.REGION);
+            if (!(regionObject instanceof String)) {
+                regionId = PersistenceConstants.DEFAULT_REGION_ID;
+                log.info("No region id provided, Hence setting the region default region");
```
Comment (mohanvive) [inline runner/components/io.siddhi.distribution.core/src/main/java/io/siddhi/distribution/core/persistence/S3PersistenceStore.java]: mention, what is the default region.

File: runner/components/io.siddhi.distribution.core/src/main/java/io/siddhi/distribution/core/persistence/S3PersistenceStore.java [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+            }
+            s3Client = clientBuilder.httpClient(httpClient).build();
+            List<Bucket> buckets;
+            try {
+                buckets = s3Client.listBuckets(ListBucketsRequest.builder().build()).buckets();
+            } catch (SdkClientException e) {
+                throw new S3PersistenceStoreException("Invalid region id provided, given region id: '" +
+                        regionId + "'.", e);
+            } catch (S3Exception e) {
+                throw new S3PersistenceStoreException("Invalid credential provided, check your user credential class" +
```
Comment (mohanvive) [inline runner/components/io.siddhi.distribution.core/src/main/java/io/siddhi/distribution/core/persistence/S3PersistenceStore.java]: How can we say that S3Exception throws only due to wrong credentials ?

File: runner/components/io.siddhi.distribution.core/src/main/java/io/siddhi/distribution/core/persistence/S3PersistenceStore.java [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+            Object bucketNameObject = configurationMap.get(PersistenceConstants.BUCKET_NAME);
+            if (!(bucketNameObject instanceof String)) {
+                throw new S3PersistenceStoreException("'bucketName' should be provided in the configuration");
+            } else {
+                this.bucketName = String.valueOf(bucketNameObject);
+                int i = Collections.binarySearch(buckets, Bucket.builder().name(this.bucketName).build(),
+                        Comparator.comparing(Bucket::name));
+                if (i < 0) {
+                    throw new S3PersistenceStoreException("Invalid bucket name provided, given bucket name: " +
+                            bucketName);
```
Comment (mohanvive) [inline runner/components/io.siddhi.distribution.core/src/main/java/io/siddhi/distribution/core/persistence/S3PersistenceStore.java]: Change the error message to provide better context.

File: runner/components/io.siddhi.distribution.core/src/main/java/io/siddhi/distribution/core/persistence/S3PersistenceStore.java [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+                }
+            }
+        } else {
+            throw new S3PersistenceStoreException("Configuration should be provided.");
```
Comment (mohanvive) [inline runner/components/io.siddhi.distribution.core/src/main/java/io/siddhi/distribution/core/persistence/S3PersistenceStore.java]: What configuration should be provided ?

File: runner/components/io.siddhi.distribution.core/src/main/java/io/siddhi/distribution/core/persistence/S3PersistenceStore.java [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+        List<S3Object> contents = listObjectsV2Response.contents();
+        for (S3Object object : contents) {
+            String key = object.key();
+            DeleteObjectRequest deleteObjectRequest =
+                    DeleteObjectRequest.builder().bucket(bucketName).key(key).build();
+            try {
+                s3Client.deleteObject(deleteObjectRequest);
+            } catch (S3Exception e) {
+                throw new CannotClearSiddhiAppStateException("Persistence state " +
+                        "object is not deleted : " + key);
```
Comment (mohanvive) [inline runner/components/io.siddhi.distribution.core/src/main/java/io/siddhi/distribution/core/persistence/S3PersistenceStore.java]: "Persisted state with id :" +key +" cannot be deleted" ?

File: runner/components/io.siddhi.distribution.core/src/main/java/io/siddhi/distribution/core/persistence/S3PersistenceStore.java [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+        ListObjectsV2Request request = ListObjectsV2Request.builder().bucket(bucketName).build();
+//        only listing out objects detail, not the object content
+        List<S3Object> objectList = s3Client.listObjectsV2(request).contents();
+        for (int i = 0; i < objectList.size() - numberOfRevisionsToSave; i++) {
+            String key = objectList.get(i).key();
+            DeleteObjectRequest deleteObjectRequest =
+                    DeleteObjectRequest.builder().bucket(bucketName).key(key).build();
+            s3Client.deleteObject(deleteObjectRequest);
+        }
+
```
Comment (mohanvive) [inline runner/components/io.siddhi.distribution.core/src/main/java/io/siddhi/distribution/core/persistence/S3PersistenceStore.java]: Remove the empty line space.

File: runner/components/io.siddhi.distribution.core/src/main/java/io/siddhi/distribution/core/persistence/S3PersistenceStore.java [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+        if (credentialProviderClassName instanceof String) {
+            log.debug("Authenticating user via the credential provider class.");
+            String className = null;
+            try {
+                className = String.valueOf(credentialProviderClassName);
+                Class credentialProviderClass = Class.forName(className);
+                return (AwsCredentialsProvider) credentialProviderClass.getDeclaredMethod("create")
+                        .invoke(credentialProviderClass);
+            } catch (IllegalAccessException | NoSuchMethodException | InvocationTargetException e) {
+                throw new S3PersistenceStoreException("Error while authenticating the user.", e);
```
Comment (mohanvive) [inline runner/components/io.siddhi.distribution.core/src/main/java/io/siddhi/distribution/core/persistence/S3PersistenceStore.java]: print the username.

File: runner/components/io.siddhi.distribution.core/src/main/java/io/siddhi/distribution/core/persistence/S3PersistenceStore.java (hunk 4 above)
Comment (sahandilshan) [reply runner/components/io.siddhi.distribution.core/src/main/java/io/siddhi/distribution/core/persistence/S3PersistenceStore.java]: `buckets = s3Client.listBuckets(ListBucketsRequest.builder().build()).buckets();` will throw exception due to 2 reasons, 
1. Invalid region id
2. Invalid credential,

When the user gives the wrong credentials, S3Client will throw S3Exception with the following message
 `The AWS Access Key Id you provided does not exist in our records`, 
There's no other way to detect if the user has given the wrong credential or not.

**Pull request merged**